File: frontend/src/modules/finance/api/cards.ts
/**
 * Typed REST client for /finance/cards.
 *
 * Wire shape matches `backend/api/openapi/finance.yaml` 1:1. All
 * domain↔wire adaptation (color/owner local metadata, brand→network
 * mapping, institution/variant derivation from the free-form name)
 * happens in the hooks layer, not here.
 */

import { apiFetch, type PageEnvelope } from "@/lib/api/client";

export type CardNetwork = "visa" | "mastercard" | "elo" | "amex";

export interface ApiCard {
  id: string;
  workspace_id: string;
  name: string;
  institution: string;
  network: CardNetwork;
  variant: string;
  last4: string;       // ^[0-9]{4}$
  limit_cents: number;
  closing_day: number; // 1..28
  due_day: number;     // 1..28
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface CreateCardRequest {
  name: string;
  institution: string;
  network: CardNetwork;
  variant?: string;
  last4: string;
  limit_cents: number;
  closing_day: number;
  due_day: number;
}

export interface UpdateCardRequest {
  name?: string;
  institution?: string;
  network?: CardNetwork;
  variant?: string;
  last4?: string;
  limit_cents?: number;
  closing_day?: number;
  due_day?: number;
}

export interface ListCardsParams {
  limit?: number;
  offset?: number;
}

function qs(params: object): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export async function listCards(p: ListCardsParams = {}): Promise<ApiCard[]> {
  const env = await apiFetch<PageEnvelope<ApiCard>>(`/finance/cards${qs(p)}`);
  return env.items;
}

export function getCard(id: string): Promise<ApiCard> {
  return apiFetch<ApiCard>(`/finance/cards/${id}`);
}

export function createCard(body: CreateCardRequest): Promise<ApiCard> {
  return apiFetch<ApiCard>(`/finance/cards`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateCard(id: string, body: UpdateCardRequest): Promise<ApiCard> {
  return apiFetch<ApiCard>(`/finance/cards/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

/** Backend uses DELETE to archive (soft-delete). */
export function archiveCard(id: string): Promise<void> {
  return apiFetch<void>(`/finance/cards/${id}`, { method: "DELETE" });
}

File: backend/internal/finance/app/cards.go
package app

import (
	"context"
	"strings"

	"github.com/google/uuid"

	"github.com/corsi/backend/internal/finance/domain"
	"github.com/corsi/backend/internal/finance/ports"
)

type CreateCardInput struct {
	WorkspaceID uuid.UUID
	Name        string
	Institution string
	Network     domain.CardNetwork
	Variant     string
	Last4       string
	LimitCents  int64
	ClosingDay  int
	DueDay      int
}

func (s *Service) CreateCard(ctx context.Context, in CreateCardInput) (*domain.Card, error) {
	c := &domain.Card{
		ID:          uuid.New(),
		WorkspaceID: in.WorkspaceID,
		Name:        strings.TrimSpace(in.Name),
		Institution: strings.TrimSpace(in.Institution),
		Network:     in.Network,
		Variant:     strings.TrimSpace(in.Variant),
		Last4:       in.Last4,
		LimitCents:  in.LimitCents,
		ClosingDay:  in.ClosingDay,
		DueDay:      in.DueDay,
	}
	if err := c.Validate(); err != nil {
		return nil, err
	}
	if err := s.repos.Cards.Create(ctx, c); err != nil {
		return nil, err
	}
	return c, nil
}

type UpdateCardInput struct {
	WorkspaceID uuid.UUID
	ID          uuid.UUID
	Name        *string
	Institution *string
	Network     *domain.CardNetwork
	Variant     *string
	Last4       *string
	LimitCents  *int64
	ClosingDay  *int
	DueDay      *int
}

func (s *Service) UpdateCard(ctx context.Context, in UpdateCardInput) (*domain.Card, error) {
	current, err := s.repos.Cards.FindByID(ctx, in.WorkspaceID, in.ID)
	if err != nil {
		return nil, err
	}
	if in.Name != nil {
		current.Name = strings.TrimSpace(*in.Name)
	}
	if in.Institution != nil {
		current.Institution = strings.TrimSpace(*in.Institution)
	}
	if in.Network != nil {
		current.Network = *in.Network
	}
	if in.Variant != nil {
		current.Variant = strings.TrimSpace(*in.Variant)
	}
	if in.Last4 != nil {
		current.Last4 = *in.Last4
	}
	if in.LimitCents != nil {
		current.LimitCents = *in.LimitCents
	}
	if in.ClosingDay != nil {
		current.ClosingDay = *in.ClosingDay
	}
	if in.DueDay != nil {
		current.DueDay = *in.DueDay
	}
	if err := current.Validate(); err != nil {
		return nil, err
	}
	if err := s.repos.Cards.Update(ctx, current); err != nil {
		return nil, err
	}
	return current, nil
}

// ArchiveCard performs a soft delete; v0.1 does not expose hard delete.
func (s *Service) ArchiveCard(ctx context.Context, workspaceID, id uuid.UUID) error {
	return s.repos.Cards.Archive(ctx, workspaceID, id)
}

func (s *Service) GetCard(ctx context.Context, workspaceID, id uuid.UUID) (*domain.Card, error) {
	return s.repos.Cards.FindByID(ctx, workspaceID, id)
}

type ListCardsInput struct {
	WorkspaceID uuid.UUID
	Limit       int
	Offset      int
}

func (s *Service) ListCards(ctx context.Context, in ListCardsInput) ([]domain.Card, error) {
	return s.repos.Cards.List(ctx, in.WorkspaceID, ports.CardFilter{Limit: in.Limit, Offset: in.Offset})
}
